Convert Studio course and library dashboard lists to React.

cahrens opened the pull request:
This PR replaces the active course, archived course, and library listings on the Studio dashboard with React code. It also fixes some a11y issues on the Studio dashboard.

[EDUCATOR-625](https://openedx.atlassian.net/browse/EDUCATOR-625)
[AC-620](https://openedx.atlassian.net/browse/AC-620)

Sandbox: https://studio-react.sandbox.edx.org/home

FYI @edx/educator-dahlia, @sstack22, @cptvitamin 

How the dashboard looks with no libraries and no archived courses (and the org optimization disabled).

![image](https://user-images.githubusercontent.com/484484/28935110-6816b488-7851-11e7-8244-0732b5e1ef17.png)

Here is an example of Archived courses working with the org optimization enabled.

![image](https://user-images.githubusercontent.com/484484/28935892-ef058f80-7853-11e7-8a3b-823dc8b5b829.png)

cahrens commented on cms/djangoapps/contentstore/views/tests/test_course_index.py:
```diff
@@ -51,55 +51,34 @@ def setUp(self):
-            # now test that url
-            outline_response = self.client.get(link.get("href"), {}, HTTP_ACCEPT='text/html')
-            self.assertEqual(outline_response.status_code, 200)
+            library_tab = parsed_html.find_class('react-library-listing')
```
Tests creating a library and then verifying that it shows up in the library listing on the studio home page: https://github.com/edx/edx-platform/blob/master/common/test/acceptance/tests/studio/test_studio_home.py#L62

Does not actually click the library link, but could add to that test.

cahrens commented on cms/djangoapps/contentstore/tests/test_contentstore.py:
```diff
@@ -1473,33 +1472,6 @@ def test_item_factory(self):
 
-    def test_course_index_view_with_course(self):
-        """Test viewing the index page with an existing course"""
-        CourseFactory.create(display_name='Robot Super Educational Course')
```
Tests creating a new course and then verifying that it shows in the course listing on the studio home page.
https://github.com/edx/edx-platform/blob/master/common/test/acceptance/tests/studio/test_studio_course_create.py#L97

Could add here clicking on the listing to make sure that it opens the course.

cahrens commented on cms/static/js/features_jsx/.eslintrc.js:
```diff
@@ -0,0 +1,7 @@
+module.exports = {
```
@andy-armstrong and @arizzitano I'm still getting jslint errors on my React file even after including this file (which I copied from the one in course_experience).

@andy-armstrong, we do plan to use the jsx extension soon, but in the very short term planned to check es2015 files into their own directory.

arizzitano replied:
@cahrens @andy-armstrong the course_experience eslintrc isn't working here because it doesn't contain any configuration for React. eslint does not parse jsx natively and [recommends extending eslint-config-airbnb](http://eslint.org/docs/user-guide/configuring#specifying-parser-options) to enable jsx support. This can be done directly within this particular eslintrc, but it's better to add it to [eslint-config-edx](https://github.com/edx/eslint-config-edx) instead so it can be used globally.

andy-armstrong replied:
Thanks for the explanation, @arizzitano! I thought we were using airbnb in our global config set, but it looks like we're only using airbnb-base. Does that not have the same benefit?

https://github.com/edx/eslint-config-edx/blob/master/packages/eslint-config-edx/.eslintrc.json

I agree that we should support JSX in our shared ruleset.

arizzitano replied:
[eslint-config-airbnb-base](https://www.npmjs.com/package/eslint-config-airbnb-base) doesn't contain any rules for linting jsx syntax. [eslint-config-airbnb](https://www.npmjs.com/package/eslint-config-airbnb-base) extends eslint-config-airbnb-base but also includes rules for linting jsx. Brian used eslint-config-airbnb-base instead of eslint-config-airbnb because we had no need for the jsx linting at the time that he implemented it, but now we will need to switch to enable full support.

arizzitano commented:
jenkins run bokchoy

cahrens commented on cms/templates/index.html:
```diff
@@ -640,4 +552,25 @@ <h3 class="title title-3">${_('Need help?')}</h3>
+    );
+    var archivedCourses = ${sorted(archived_courses, key=lambda s: s['display_name'].lower() if s['display_name'] is not None else '') | n, dump_js_escaped_json};
+
+    if (archivedCourses.length > 0) {
```
TODO: I can remove this given the check in the jsx file.

cahrens commented on cms/djangoapps/contentstore/tests/test_contentstore.py:
```diff
@@ -1473,33 +1472,6 @@ def test_item_factory(self):
-            html=True
-        )
-
-    def test_course_index_view_xss(self):
```
Since the lists are now client-side rendered, these tests need to be removed. I tested with XSS injections and had no problems with the React code (and note that we already have bok choy test coverage for XSS for the dashboard page).

arizzitano replied:
Yeah! React has XSS protection built in: https://facebook.github.io/react/docs/introducing-jsx.html

cahrens commented on cms/djangoapps/contentstore/tests/test_course_listing.py:
```diff
@@ -88,30 +86,6 @@ def tearDown(self):
         self.client.logout()
         ModuleStoreTestCase.tearDown(self)
 
-    def test_course_listing_is_escaped(self):
```
Another XSS server-side test.

cahrens commented on cms/djangoapps/contentstore/views/tests/test_course_index.py:
```diff
@@ -402,31 +381,8 @@ def check_index_page(self, separate_archived_courses, org):
         parsed_html = lxml.html.fromstring(index_response.content)
         course_tab = parsed_html.find_class('courses')
         self.assertEqual(len(course_tab), 1)
-        course_links = course_tab[0].find_class('course-link')
```
I had to remove the checks about specific courses, since the listings are client-side rendered by React. However, I left whether or not the top-level "archived" tab appears, since that is still server-side rendered.

I covered which list the courses appear in with bok choy tests (below).

cahrens commented on cms/envs/common.py:
```diff
@@ -252,6 +252,10 @@
+
+    # Whether archived courses (courses with end dates in the past) should be
+    # shown in Studio in a separate list.
+    'ENABLE_SEPARATE_ARCHIVED_COURSES': True
```
When the OSPR merged (#15431), the feature flag was not enabled for Studio. @sstack22 does want it enabled though.

cahrens commented on cms/templates/index.html:
```diff
@@ -73,7 +76,7 @@ <h3 class="title">${_("Create a New Course")}</h3>
                   <span class="tip tip-error is-hiding" id="tip-error-new-course-name"></span>
                 </li>
-                <li class="field text required" id="field-organization">
+                <li class="field text required">
```
Remove duplicated IDs (which weren't used anyway).

cahrens commented on common/test/acceptance/pages/studio/index.py:
```diff
@@ -205,12 +205,13 @@ def select_item_in_autocomplete_widget(self, item_text):
         self.q(css='.ui-autocomplete .ui-menu-item a').filter(lambda el: el.text == item_text)[0].click()
 
-    def list_courses(self):
+    def list_courses(self, archived=False):
```
Because of the default value, existing usages of this method will end up verifying (indirectly) that active courses are on the active list instead of the archived list.

cahrens commented:
jenkins run lettuce

cahrens commented on cms/templates/index.html:
```diff
@@ -473,68 +442,11 @@ <h4 class="title">${_('Your Course Creator Request Status:')}</h4>
-          %endfor
-        </ul>
-      </div>
+      %if len(libraries) > 0 or optimization_enabled:
```
The "do you want to create your first library" messaging shouldn't show if the org optimization is enabled (it was an oversight when the feature was extended to include libraries). Here is how it looks now when org is not specified:

![image](https://user-images.githubusercontent.com/484484/28935420-8987a39c-7852-11e7-9f1e-a0b36426a160.png)

cahrens commented on cms/static/js/features_jsx/studio/index.jsx:
```diff
@@ -0,0 +1,102 @@
+                                    <li className="action action-rerun">
+                                        <a href={item.rerun_link}
+                                           className="button rerun-button"
+                                           aria-describedby={"title-"+idBase+"-"+i}
```
Note the usage of aria-describedby to address https://openedx.atlassian.net/browse/AC-620.

cahrens replied:
@cptvitamin Can you take a quick look at the sandbox and make sure you are happy with this?
https://studio-react.sandbox.edx.org/home

cahrens commented:
Ok, @arizzitano, finally a green build! Please review when you get a chance.

arizzitano commented on cms/static/js/features_jsx/studio/index.jsx:
```diff
@@ -0,0 +1,102 @@
+import React, {Component} from 'react';
```
Actually, I don't think eslint is running against this file at all, since it doesn't yet know to look at `.jsx` files. ESLint determines which filetypes to lint based on `--ext` parameters [passed on the command line](http://eslint.org/docs/user-guide/configuring#specifying-file-extensions-to-lint) which fortunately is pretty straightforward to fix. Within our `run_eslint` quality task, just add `--ext .js --ext .jsx` before `--format-compact` on this line: https://github.com/edx/edx-platform/blob/7bc87a5f5672e65a01034ed850aeb43f8813ef3a/pavelib/quality.py#L291.

cahrens commented:
jenkins run quality

cahrens commented:
jenkins run quality

cahrens commented:
Jenkins run quality

cahrens commented:
jenkins run quality

cahrens commented:
jenkins run quality

cahrens commented:
jenkins run quality

cahrens commented:
@arizzitano I have fixed all the quality issues in the file and decreased the number of jslint errors to the current number on this branch. I've also updated the sandbox. Please continue reviewing when you are able.

cahrens commented on cms/static/js/features_jsx/studio/index.jsx:
```diff
@@ -1,104 +1,131 @@
-class CourseOrLibraryListing extends React.Component {
+function CourseOrLibraryListing(props) {
+  /* global gettext */
+  /* eslint react/no-array-index-key: 0 */
```
@arizzitano I added this because these lists are not dynamic. Therefore there is no risk of elements being added/removed in the wrong location in the list, and it is safe to just use the index as the key.

arizzitano replied:
Yeah, honestly I kind of hate `react/no-array-index-key` and was really tempted to disable it globally in `eslint-config-edx`. I think it's okay to disable inline for now, but I'd move these two lines up to the very top of the file, the same way we do with pylint ignore rules.

arizzitano commented:
@cahrens awesome work pulling this all together and pushing the frontend towards modernity. Here's to the beginning of React in edx-platform 🥂 🚀🎆

cahrens commented:
jenkins run python
